linkchecker.go

```diff
@@ -0,0 +1,437 @@
+	l.Wg.Add(1)
+	l.GetResult(site, referringSite)
+	l.Wg.Wait()
+
+	resp, err := l.GetResponse(site)
+	if err != nil {
+		fmt.Fprint(l.errorLog, err)
+	}
+
+	u, err := url.Parse(site)
+	if err != nil {
+		fmt.Fprint(l.errorLog, err)
+	}
+
+	// crawl pages and check links within domain
+	// and only those not in the CheckLink.List map
+	if u.Host == l.Domain && !l.IsCrawled(site) {
+
+		// add site to CheckLink.List map to handle
+		// link back (e.g pageA links pageB which links to pageA)
+		// add site in if statement to allow the top level page to
+		// be crawled.
+		l.AddSite(site)
+
+		// generate of list of links on page
+		links, err := l.ParseBody(resp.Body)
+		if err != nil {
+			fmt.Fprintf(l.errorLog, "unable to generate site list, %s", err)
+		}
+
+		for _, link := range links {
+
+			if !l.IsCrawled(link) {
+				l.Wg.Add(1)
+				go l.GetResult(link, site)
+				l.Wg.Wait()
+			}
+
+		}
+
+	}
+
+}
+
+func (l *LinkChecker) GetResult(site string, referringSite string) {
+
+	defer l.Wg.Done()
+	result := Result{
+		Url:           site,
+		ReferringSite: referringSite,
+	}
+
+	switch IsHttpTypeLink(site) {
+	case true:
+		resp, err := l.GetResponse(site)
+		if err != nil {
+			result.Error = err
+		}
+		result.ResponseCode = resp.StatusCode
+
+		_, err = url.Parse(site)
+		if err != nil {
+			result.Error = err
+		}
+
+	case false:
+		result.Error = fmt.Errorf("unable to check non http/https links: %s", site)
+	}
+	l.Results <- result
+
+}
+
+func (l *LinkChecker) GetResponse(site string) (*http.Response, error) {
+
+	ctx := context.Background()
+	err := l.Ratelimiter.Wait(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	_, err = l.IsHeaderAvailable(site)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := l.HTTPClient.Get(site)
+	if err != nil {
+
+		return resp, err
+	}
+
+	return resp, nil
+}
+
+func (l *LinkChecker) ParseBody(body io.Reader) ([]string, error) {
+
+	sites := []string{}
+	doc, err := htmlquery.Parse(body)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse body, check if a valid io.Reader is being sent, %s", err)
+	}
+
+	list := htmlquery.Find(doc, "//a/@href")
+
+	for _, n := range list {
+		url := htmlquery.InnerText(n)
+
+		if IsHttpTypeLink(url) {
+			url, err = l.CanonicaliseChildUrl(url)
+			if err != nil {
+				fmt.Fprintf(l.errorLog, "unable to canonicalise url: %s, %s", url, err)
+			}
+		}
+		sites = append(sites, url)
+
+	}
+
+	return sites, nil
+}
+
+func IsHttpTypeLink(link string) bool {
+
+	return !strings.HasPrefix(strings.ToLower(link), "mailto:") && !strings.HasPrefix(strings.ToLower(link), "ftp:")
+}
+
+func (l *LinkChecker) IsHeaderAvailable(site string) (bool, error) {
+
+	resp, err := l.HTTPClient.Head(site)
+	if err != nil {
+		return false, err
+	}
+
+	defer resp.Body.Close()
+
+	return true, nil
+}
+
+type CheckLink struct {
+	Locker sync.Mutex
+	List   map[string]bool
+}
+
+func (l *LinkChecker) IsCrawled(site string) bool {
+
+	result := false
+	if l.CheckLink.List[site] {
```

Comment:
Don't we need to get the mutex? (A read lock would be OK, so we could use a `sync.RWMutex` and call `RLock`.)